daemon/src/order/maker.rs

```diff
@@ -0,0 +1,385 @@
+use model::Role;
+use std::collections::HashMap;
+use std::fmt;
+use std::time::Duration;
+use tokio_extras::FutureExt;
+use xtra::prelude::MessageChannel;
+use xtra_libp2p::NewInboundSubstream;
+use xtra_libp2p::Substream;
+use xtra_productivity::xtra_productivity;
+use xtras::SendAsyncSafe;
+
+const ORDER_TIMEOUT: Duration = Duration::from_secs(5);
+
+pub struct Actor {
+    executor: command::Executor,
+    oracle_pk: XOnlyPublicKey,
+    get_announcement:
+        MessageChannel<oracle::GetAnnouncement, Result<olivia::Announcement, NoAnnouncement>>,
+    build_party_params: MessageChannel<wallet::BuildPartyParams, Result<PartyParams>>,
+    sign: MessageChannel<wallet::Sign, Result<PartiallySignedTransaction>>,
+    projection: xtra::Address<projection::Actor>,
+    n_payouts: usize,
+    decision_senders: HashMap<OrderId, oneshot::Sender<protocol::Decision>>,
+    db: sqlite_db::Connection,
+    latest_offers: MessageChannel<xtra_libp2p_offer::maker::GetLatestOffers, Option<MakerOffers>>,
+}
+
+impl Actor {
+    pub fn new(
+        n_payouts: usize,
+        oracle_pk: XOnlyPublicKey,
+        get_announcement: MessageChannel<
+            oracle::GetAnnouncement,
+            Result<olivia::Announcement, NoAnnouncement>,
+        >,
+        (db, process_manager): (sqlite_db::Connection, xtra::Address<process_manager::Actor>),
+        (build_party_params, sign): (
+            MessageChannel<wallet::BuildPartyParams, Result<PartyParams>>,
+            MessageChannel<wallet::Sign, Result<PartiallySignedTransaction>>,
+        ),
+        projection: xtra::Address<projection::Actor>,
+        latest_offers: MessageChannel<
+            xtra_libp2p_offer::maker::GetLatestOffers,
+            Option<MakerOffers>,
+        >,
+    ) -> Self {
+        Self {
+            executor: command::Executor::new(db.clone(), process_manager),
+            oracle_pk,
+            get_announcement,
+            build_party_params,
+            sign,
+            projection,
+            n_payouts,
+            decision_senders: HashMap::default(),
+            db,
+            latest_offers,
+        }
+    }
+
+    async fn await_order(
+        &mut self,
+        framed: &mut Framed<Substream, JsonCodec<MakerMessage, TakerMessage>>,
+    ) -> Result<TakerMessage> {
+        let order = framed
+            .next()
+            .timeout(ORDER_TIMEOUT, || tracing::debug_span!("receive order"))
+            .await
+            .context("Timout when waiting for order")?
+            .context("Stream terminated")?
+            .context("Unable to decode order")?;
+
+        Ok(order)
+    }
+
+    async fn pick_offer(&self, offer_id: OfferId) -> Result<Offer> {
+        // TODO: Evaluate if this can be a problem during sending new offers to all connected peers
+        // -> Should be fine because the offer is already outdated by then and we should actually
+        // wait until finished?
+        // -> Cannot use send_async_next here - is that expected?
```

Comment:
yes, it's expected that we can't use `send_async_next` here; we are awaiting the completion of the handler as we need the result back :)

note: we could use `send(msg).priority(NUMBER)` if we wanted to utilise higher priority to ensure that this protocol does go smoothly (at the risk of actually getting an outdated offer, so I'm not sure whether it's worth it)

Reply:
I think we can "solve" this by moving this decision into the automation, but we have to be aware that we will have to manage the state there then. I will see to write down a ticket for this and action it.
For now I think this solution is good enough, and I don't expect to run into too much trouble, because sending to each peer is in a separate task, so the offer actor should not be blocked for too long.